src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -433,50 +452,70 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
         return CompletableFuture.completedFuture(0);
     }
 
-    private synchronized void spoolMessage()  {
-        if (spoolingFuture.get() == null || spoolingFuture.get().isCancelled()) {
-            spoolingFuture.set(ses.scheduleWithFixedDelay(this::spoolTask, 0, 5, TimeUnit.SECONDS));
+    @SuppressWarnings("PMD.AvoidCatchingThrowable")
+    protected CompletableFuture<Integer> publishSingleSpoolerMessage() throws InterruptedException {
+        final long id = spool.popId();
+        PublishRequest request = spool.getMessageById(id);
+        if (request == null) {
+            return null;
+        }
+        MqttMessage m = new MqttMessage(request.getTopic(), request.getPayload());
+        try {
+            return getConnection(false).publish(m, request.getQos(), request.isRetain())
+                    .whenComplete((packetId, throwable) -> {
+                        // packetId is the SDK assigned ID. Ignore this and instead use the spooler ID
+                        if (throwable == null) {
+                            spool.removeMessageById(id);
+                            logger.atDebug().kv("id", id).kv("topic", request.getTopic())
+                                    .log("Successfully published message");
+                        } else {
+                            spool.addId(id);
```

Comment:
Here and line477. Have you thought about how to handle a "poison pill" message? For example, if the message size is larger than 128KB and rejected by IoT, do we end up retrying infinitely and blocking all the other messages?

Reply:
Right now, we always retry until it succeed. For the message length, it would be better to reject it when it is trying to insert the queue if that is too long. But for other non-retryable message, it may block the queue. Do we need to add some retry logic here?

Reply:
I think it is better to create new class called SpoolMessage which has three field, (1) id, (2)publish request, (3) retriedTime. The decision on whether to add the message back could be made in this block of code based on the 'retriedTimes'. In this way, we could prevent other kind of non-retryable exception

Or add it as TODO and make changes in the next release?

Reply:
Ideally we should be able to distinguish whether the exception is retryable or not. If it's not retryable, then the message can be discarded and log an error.